proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+   synonym, the synonym is expanded. It is an error if that location does not allow wildcards.
+   The wildcard is understood to stand for just one type shared among all the expansions.
+
+Effects
+^^^^^^^
+
+1. We can now bind local type synonyms, avoiding the need to do so via pattern or result
+   signatures.
+
+#. One challenge is how to present these local synonyms in error messages. It might be
+   best to aggressively expand (unlike top-level type synonyms), especially because these
+   local synonyms might refer to other local type variables that are in scope. As we gain
+   experience with this new form, we can refine their appearance in error messages.
+
+#. Note that this proposal does *not* allow for top-level lower-case type synonyms. There
+   is nothing stopping us from doing so, but it would seem to violate expectations of Haskellers
+   and would be the first instance of a lower-case type variable being in scope at the top level.
+
+``let`` in patterns
+~~~~~~~~~~~~~~~~~~~
+
+This part of this proposal allows introducing a ``let``\ -binding in a pattern.
+The bound variable(s) scope over the same region of code as the pattern-bound
+variables do.
+
+The syntax for this feature is a bit awkward, and so this proposal presents
+two alternatives for discussion.
+
+Motivation
+^^^^^^^^^^
+
+1. A careful reader will note that allowing ``let`` for `type synonyms <#type-let>`_
+   does not, by itself, replace a binding such as that in ``f (True :: b) = ...``
+   because the current form binds the type variable in the pattern. This part
+   of the proposal closes this gap. See `examples <#let-in-pattern-example>`_ below.
+
+#. Though admittedly a weakish motivation, there is currently no way to share
+   expressions used in common in multiple view patterns. See `examples <#let-in-pattern-example>`_
+   below.
+
+Specification Alternative 1
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. With ``-XExtendedLet``, add a new form of pattern as follows::
+
+     pat → 'let' decls 'in' pat
+
+#. Any entites bound in ``decls`` scope over the same region of the program
+   that pattern-bound variables scope over, with the addition of the ``decls``
+   themselves (that is, the declarations can be recursive).
+
+Specification Alternative 2
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+0. **Background**. Here are some productions from the `Haskell 2010 Report`_::
+
+     funlhs → var apat {apat}
+            | pat varop pat
+            | '(' funlhs ')' apat {apat}
+
+     apat → gcon
+          | literal
+          | …
+
+     lexp → '\' apat1 … apatn '->' exp   (n ≧ 1)
+          | …
+
+     lpat → apat
+          | '-' (integer|float)
+          | gcon apat1 … apatn
+
+   (Recall that braces mean "0 or more".)
+
+1. With ``-XExtendedLet``, change the grammar to be ::
+
+     funlhs → var apats1
+            | pat varop pat
+            | '(' funlhs ')' apats
+
+     apats1 → apat
+            | apat apats
+            | '(' 'let' decls ')' apats
```

Comment:
Is this syntax supposed to suggest symmetry with the `funlhs` production? If so, why is it, they don't seem to bear any relationship to my eyes.

It does resemble the syntax for the `do` notation though. How about just making `let decls` be a `pat`. Then the parentheses come from `apat → '(' 'pat ')'`.

Reply:
No resemblance to `funlhs` intended. Making `let decls` be a `pat` means it would be allowed anywhere... except that it would match against nothing. It really only makes sense in a `pats` (plural), where it can be ignored during matching.

I personally don't see a need or place for this in the proposal text. If you think it belongs there, can you suggest a spot?